src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -113,46 +110,47 @@ typesOf bindingGroupType moduleName vals = withFreshSubstitution $ do
           let solved = foldMap (S.fromList . fdDetermined) typeClassDependencies
           let constraintTypeVars = nub . foldMap (unknownsInType . fst) . filter ((`notElem` solved) . snd) $ zip (constraintArgs con) [0..]
           when (any (`notElem` unsolvedTypeVars) constraintTypeVars) $ do
-            nextVar <- peek
-            throwError . onErrorMessages (replaceTypes (not shouldGeneralize) currentSubst nextVar) . errorMessage $ NoInstanceFound con
+            checkState <- get
+            throwError . onErrorMessages (replaceTypes Nothing currentSubst checkState) . errorMessage $ NoInstanceFound con
 
       -- Check skolem variables did not escape their scope
       skolemEscapeCheck val'
       -- Check rows do not contain duplicate labels
       checkDuplicateLabels val'
-      return (ident, (foldr (Abs . Left . fst) val' unsolved, generalized))
+      return ((ident, (foldr (Abs . Left . (\(x, _, _) -> x)) val' unsolved, generalized)), unsolved)
 
     -- Show warnings here, since types in wildcards might have been solved during
     -- instance resolution (by functional dependencies).
     finalSubst <- gets checkSubstitution
+    checkState <- get
+    forM_ tys $ \(shouldGeneralize, ((_, (_, _)), w)) -> do
+      escalateWarningWhen isHoleError . tell . onErrorMessages (replaceTypes (guard shouldGeneralize $> concatMap snd inferred) finalSubst checkState) $ w
 
-    -- TODO: We should only do type search for Typed Holes which we inferred
-    -- without generalizing type class constraints
-    -- eg. foldMap ?x [1, 2, 3] finds `negate` right now, which is wrong
-
-    nextVar <- peek
-    escalateWarningWhen isHoleError . tell . onErrorMessages (replaceTypes True finalSubst nextVar) $ w
-
-    return inferred
+    return (map fst inferred)
   where
-    replaceTypes shouldRunTypeSearch subst nextVar =
-      (if shouldRunTypeSearch then runTypeSearch else id)
-      . onTypesInErrorMessage (substituteType subst)
+    replaceTypes
+      :: Maybe [(Ident, InstanceContext, Constraint)]
+      -> Substitution
+      -> CheckState
+      -> ErrorMessage
+      -> ErrorMessage
+    replaceTypes cons subst st =
```

Comment:
Maybe we should pull this out of the where clause and give it a few haddocks? It doesn't depend on any of the arguments of `typesOf` and is complicated enough to need its own signature.